### ISSUE TITLE
Introduce an cutoff when searching for relevant memories 

### ISSUE THREAD

**Issue by syntex01** [0arm/emergent#7]
The function that searches a kB article that is similar to the prompt, could also include an variable that defines a minimum distance in the embedded space so that any article is returned. Sometimes and even very often, the closest article might have absolutely no relationship with the topic. This will greatly reduce token use. Not only when providing kb articles but also when we want to identify how many articles to update. This could be done dynamically instead of trying to update the closest three.

**Comment by kyb3r** [comment]
threshold of > 0.6 cosine_similarity should be ok, very very different texts usually have similarity below 0.5, similar texts are usually 0.75 upwards

**Comment by syntex01** [comment]
Yes and we can also just experiment to find a value that works well.

**Comment by haukepribnow** [comment]
Out of curiosity: How sure are you that embeddings & cosine similarity will really be enough for successful knowledge linking at a large scale? I've just posted a long post with several questions about this in the discussion section of David's HMCS repository (https://github.com/daveshap/HierarchicalMemoryConsolidationSystem/discussions/4) - feel free to chime in there too if you like.

**Comment by kyb3r** [comment]
Right now we don’t use only cosine similarity. We also use chatgpt to classify whether or not a knowledge base article is relevant to a roll up summary.

So cosine similarity is kinda of like a filter, but cosine similarity is not always accurate, so using an LLM as a function to make a classification is what we are doing in addition to that

**Comment by haukepribnow** [comment]
Understood - thanks a lot for your response.

With my question, I was referring less to roll-up summaries but more to the "recall" mechanism. According to the following code... https://github.com/kyb3r/emergent/blob/b59165d0c1880df6f9e62b968c4625d3098150ec/emergent/memory.py#L261-L273 ...a cosine similarity comparison is used to determine the (one) most likely KB article that may be relevant in answering user input.

I was wondering on your thoughts on that part: How do you plan to further develop this "recall" (/ "lookup") mechanism? How do you expect it to perform at a scale? And how do you plan on handling the situation when one KB article happens to have a more similar embedding vector to the user input than another, despite the found KB article being the "wrong" one for the respective situation?

**Comment by syntex01** [comment]
So one thing we thought about was introducing an internal  dialogue of the chat agent. In this internal monologue the agent will decide if additional information is needed. And will then state what the information should be about. We will then search for the closest KB articles in embedded space. The agent will be provided with the headings of the n closest articles and will then decide which are in fact relevant. This method has the drawback of using up quite a few tokens, but I don't see any other way of addressing the problems you also identified.

**Comment by haukepribnow** [comment]
Ah, that's an interesting approach. Thanks a lot for sharing.

**Comment by kyb3r** [comment]
![IMG_0245](https://user-images.githubusercontent.com/29310830/228185365-4ced5f51-196f-41c8-a0ad-69dc431e9710.jpeg)

I read this conversation and its an interesting idea. 

The plan is to give the agent access to tools similar to chatgpt plugins. So it will autonomously choose when it needs to get a relevant memory. Hypothetical document embeddings could be a possible way to make the query that agent makes more accurate.
